core/src/connection_reuse.rs

```diff
@@ -236,42 +391,304 @@ where
+                            true
+                        }
+                        Err(err) => {
+                            // If we get an error while opening a substream, we decide to ignore it
+                            // and open a new muxer.
+                            // If opening the muxer produces an error, *then* we will return it.
+                            debug!(
+                                "Error while opening outgoing substream to {}: {:?}",
+                                self.addr, err
+                            );
+                            true
+                        }
+                    }
+                }
+                _ => false,
+            };
+
+            let is_empty = {
+                let mut conns = self.shared.connections.lock();
+                if should_kill_existing_muxer {
+                    conns.remove(&self.addr);
+                    true
+                } else {
+                    let item = conns.get(&self.addr);
+
+                    // error'ed, quit early
+                    if let Some(PeerState::Errored(err)) = item {
+                        trace!(
+                            "Existing new connection to {} errored earlier: {:?}",
+                            self.addr,
+                            err
+                        );
+                        let io_err = IoError::new(err.kind(), err.to_string());
+                        return Err(io_err);
+                    };
+
+                    item.is_none()
+                }
+            };
+
+            if is_empty {
+                // dial new and return early
+                let state = match self.shared.transport.clone().dial(self.addr.clone()) {
+                    Ok(future) => {
+                        trace!("Opened new connection to {:?}", self.addr);
+                        let future = future.and_then(|(out, addr)| addr.map(move |a| (out, a)));
+                        let future = Box::new(future);
+                        let connection_id = self.shared.gen_next_connection_id();
+                        PeerState::Pending {
+                            future,
+                            notify: Default::default(),
+                            connection_id,
+                        }
+                    }
+                    Err(_) => {
+                        trace!(
+                            "Failed to open connection to {:?}, multiaddr not supported",
+                            self.addr
+                        );
+                        let err =
+                            IoError::new(IoErrorKind::ConnectionRefused, "multiaddr not supported");
+                        PeerState::Errored(err)
+                    }
+                };
+
+                self.shared.insert_connection(self.addr.clone(), state);
+                return Ok(Async::NotReady);
```

Comment:
Here we have just stored the dial future, but we have not polled it yet. By just returning `NotReady` we may not be woken up again, no?